test/net/sourceforge/kolmafia/request/RequestTestBase.java

```diff
@@ -32,4 +86,175 @@ protected void expectSuccess(GenericRequest spy, String responseText) {
+            }
+          }
+          return true;
+        });
+  }
+
+  /**
+   * Is true if the base of the URL matches {@code path}. The "base" is the part coming before the
+   * {@code ?} (if any), e.g., the base of {@code inventory.php?which=2} is {@code inventory.php}
+   */
+  protected PathMatcher hasPath(final String path) {
+    return new PathMatcher(path);
+  }
+
+  /**
+   * Is true if one of the URL's parameters has a key matching {@code param}, and a value matching
+   * {@code value}. The parameters are the part coming after the {@code ?} (if any), separated by
+   * {@code &}s.
+   *
+   * <p>For example, in {@code inventory.php?which=2&foo&bar=&pwd=abc}, the parameters are {@code
+   * which=2}, {@code foo}, {@code bar=} and {@code pwd=abc}.
+   *
+   * <p>Parameters are split into a key and a value, the key being what comes before the {@code =}
+   * (if any), and the value being what's after (again, if any).
+   *
+   * <p>If {@code value} is {@code null}, the parameter's value is allowed to be anything.
+   */
+  protected ParamMatcher hasParam(final String param, final String value) {
+    return new ParamMatcher(param, value);
+  }
+
+  /**
+   * Is true if one of the URL's parameters has a key matching {@code param}, and a value matching
+   * {@code value}. The parameters are the part coming after the {@code ?} (if any), separated by
+   * {@code &}s.
+   *
+   * <p>For example, in {@code inventory.php?which=2&foo&bar=&pwd=abc}, the parameters are {@code
+   * which=2}, {@code foo}, {@code bar=} and {@code pwd=abc}.
+   *
+   * <p>Parameters are split into a key and a value, the key being what comes before the {@code =}
+   * (if any), and the value being what's after (again, if any).
+   */
+  protected ParamMatcher hasParam(final String param, final Pattern valueMatcher) {
+    return new ParamMatcher(param, valueMatcher);
+  }
+
+  private void prepareResponse(final String response, final ArgumentMatcher<URL> matcher) {
+    try {
+      when(factory.openConnection(argThat(matcher))).thenReturn(new FakeConnection(response));
+    } catch (IOException e) {
+    }
+  }
+
+  private class FakeConnection extends HttpURLConnection {
+    private final String responseText;
+
+    private FakeConnection(final String responseText) {
+      super(null);
+      this.responseText = responseText;
+      this.responseCode = 200;
+    }
+
+    @Override
+    public InputStream getInputStream() throws IOException {
+      return new ByteArrayInputStream(this.responseText.getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Override
+    public void disconnect() {}
+
+    @Override
+    public boolean usingProxy() {
+      return false;
+    }
+
+    @Override
+    public void connect() throws IOException {}
+  }
+
+  private class SectionedUrl {
+    private final String path;
+    private final Map<String, String> params = new HashMap<>();
+
+    private SectionedUrl(final URL url) {
+      final String fullPath = url.getPath();
```

Comment:
This won't work.

getPath() only returns the part of the path prior to the query and/or fragment.